crates/oxc_index/src/idxslice.rs

```diff
@@ -0,0 +1,847 @@
+#[allow(clippy::wildcard_imports)]
+use super::*;
+
+/// A slice that only accepts indices of a specific type. Note that the intended
+/// usage is as `IndexSlice<I, [T]>`.
+///
+/// This is a thin wrapper around a `[T]`, to the point where the backing  is a
+/// public property (called `raw`). This is in part because I know this API is
+/// not a complete mirror of Vec's (patches welcome). In the worst case, you can
+/// always do what you need to the slice itself.
+///
+/// ## Some notes on the APIs
+///
+/// - Most of the Slice APIs are present.
+///     - Any that aren't can be trivially accessed on the underlying `raw`
+///       field, which is public.
+///
+/// - Apis that take or return usizes referring to the positions of items were
+///   replaced with ones that take Idx.
+///
+/// - Apis that take `R: RangeBounds<usize>` take an
+///   [`IdxRangeBounds<I>`][IdxRangeBounds], which is basically a
+///   `RangeBounds<I>`.
+/// - Apis that take `SliceIndex<usize>` take an
+///   [`IdxSliceIndex<I>`][IdxSliceIndex], which is basically a `SliceIndex<I>`.
+///
+/// - Most iterator functions where `the_iter().enumerate()` would refer to
+///   indices have been given `_enumerated` variants. E.g.
+///   [`IndexSlice::iter_enumerated`], etc. This is because
+///   `v.iter().enumerate()` would be `(usize, &T)`, but you want `(I, &T)`.
+///
+/// The following extensions are added:
+///
+/// - [`IndexSlice::indices`]: an Iterator over the indices of type `I`.
+/// - Various `enumerated` iterators mentioned earlier
+/// - [`IndexSlice::position`], [`IndexSlice::rposition`] as
+///   `self.iter().position()` will return a `Option<usize>`
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct IndexSlice<I: Idx, T: ?Sized> {
+    _marker: PhantomData<fn(&I)>,
+    pub raw: T,
+}
+
+// TODO: we may want to remove this somehow. Won't work in upcoming rust versions.
+#[allow(unsafe_code, suspicious_auto_trait_impls)]
+// SAFETY: Whether `IndexSlice` is `Send` depends only on the data,
+// not the phantom data.
+unsafe impl<I: Idx, T> Send for IndexSlice<I, [T]> where T: Send {}
```

Comment:
This will break in the short term future. https://github.com/rust-lang/rust/issues/93367

Reply:
I'll create an issue for it after the merge so we can keep track of it.